source/opt/merge_return_pass.cpp

```diff
@@ -239,8 +243,22 @@ void MergeReturnPass::CreatePhiNodesForInst(BasicBlock* merge_block,
   if (inst.result_id() != 0) {
     std::vector<Instruction*> users_to_update;
     context()->get_def_use_mgr()->ForEachUser(
-        &inst, [&users_to_update, &dom_tree, inst_bb, this](Instruction* user) {
-          BasicBlock* user_bb = context()->get_instr_block(user);
+        &inst,
+        [&users_to_update, &dom_tree, &inst, inst_bb, this](Instruction* user) {
+          BasicBlock* user_bb = nullptr;
+          if (user->opcode() != SpvOpPhi) {
+            user_bb = context()->get_instr_block(user);
+          } else {
+            // For OpPhi, the use should be considered to be in the predecessor.
```

Comment:
But not just in the predecessor, right?  A Phi operand could come from any arbitrary CFG path that reaches this basic block.  I'm thinking a pseudocode like this:

```
BB0:
%4 = ...
BB1:
....
if (cond)
THEN_BB:
   ...
ELSE_BB:
   ...
   %5 = ...
ENDIF_BB:
%10 = OpPhi(%4@BB0, %5@ELSE_BB)
```

`BB0` is not a predecessor of `ENDIF_BB`, but it reaches it.

Or am I just confused by the phrasing of the comment?  The code seems to check just predecessors.

Reply:
I don't think your example is correct.  An OpPhi will mention the immediate predecessors of the node.  Your example should be:

```
BB0:
%4 = ...
BB1:
....
if (cond)
THEN_BB:
   ...
ELSE_BB:
   ...
   %5 = ...
ENDIF_BB:
%10 = OpPhi(%4, THEN_BB, %5, ELSE_BB)
```

We will pretend that the use of `%4` took place in `THEN_BB`, so we need to make sure that the def of `%4` dominates `THEN_BB`, not `ENDIF_BB`.

Reply:
> I don't think your example is correct. An OpPhi will mention the immediate predecessors of the node. Your example should be:
> 
> ```
> BB0:
> %4 = ...
> BB1:
> ....
> if (cond)
> THEN_BB:
>    ...
> ELSE_BB:
>    ...
>    %5 = ...
> ENDIF_BB:
> %10 = OpPhi(%4, THEN_BB, %5, ELSE_BB)
> ```

Sorry, yes.  `%4` comes from `THEN_BB`

> We will pretend that the use of `%4` took place in `THEN_BB`, so we need to make sure that the def of `%4` dominates `THEN_BB`, not `ENDIF_BB`.

Ah, that predecessor.   Now I get it.  Thanks.